File: rust/tests/failure_handler.rs
//! Tests for failure_handler.rs
//!
//! Mirrors failure_handler test coverage from the JS test suite.

use start_command::failure_handler::{
    can_create_issue, handle_failure, is_gh_authenticated, is_gh_upload_log_available,
    parse_git_url, Config,
};

mod parse_git_url_tests {
    use super::*;

    #[test]
    fn should_parse_https_url() {
        let info = parse_git_url("https://github.com/owner/repo").unwrap();
        assert_eq!(info.owner, "owner");
        assert_eq!(info.repo, "repo");
        assert_eq!(info.url, "https://github.com/owner/repo");
    }

    #[test]
    fn should_parse_ssh_url() {
        let info = parse_git_url("git@github.com:owner/repo.git").unwrap();
        assert_eq!(info.owner, "owner");
        assert_eq!(info.repo, "repo");
    }

    #[test]
    fn should_parse_git_at_format() {
        let info = parse_git_url("git@github.com:myorg/myrepo").unwrap();
        assert_eq!(info.owner, "myorg");
        assert_eq!(info.repo, "myrepo");
    }

    #[test]
    fn should_return_none_for_empty_string() {
        assert!(parse_git_url("").is_none());
    }

    #[test]
    fn should_return_none_for_invalid_url() {
        assert!(parse_git_url("not a url").is_none());
    }

    #[test]
    fn should_return_none_for_non_github_url() {
        assert!(parse_git_url("https://gitlab.com/owner/repo").is_none());
    }

    #[test]
    fn should_strip_git_suffix_from_repo() {
        let info = parse_git_url("https://github.com/owner/repo.git").unwrap();
        assert_eq!(info.repo, "repo");
        assert!(!info.repo.ends_with(".git"));
    }

    #[test]
    fn should_normalize_url_to_https() {
        let info = parse_git_url("git@github.com:owner/myrepo.git").unwrap();
        assert!(info.url.starts_with("https://"));
        assert!(info.url.contains("github.com"));
    }

    #[test]
    fn should_parse_npm_style_git_url() {
        // npm registry returns git+https:// URLs
        let info = parse_git_url("git+https://github.com/link-foundation/start.git");
        // Should either parse or return None (regex matches github.com)
        if let Some(i) = info {
            assert_eq!(i.owner, "link-foundation");
            assert_eq!(i.repo, "start");
        }
    }
}

mod config_tests {
    use super::*;

    #[test]
    fn should_have_correct_default_values() {
        let config = Config::default();
        assert!(!config.disable_auto_issue);
        assert!(!config.disable_log_upload);
        assert!(!config.verbose);
    }

    #[test]
    fn should_allow_setting_disable_auto_issue() {
        let config = Config {
            disable_auto_issue: true,
            ..Config::default()
        };
        assert!(config.disable_auto_issue);
    }

    #[test]
    fn should_allow_setting_disable_log_upload() {
        let config = Config {
            disable_log_upload: true,
            ..Config::default()
        };
        assert!(config.disable_log_upload);
    }

    #[test]
    fn should_allow_setting_verbose() {
        let config = Config {
            verbose: true,
            ..Config::default()
        };
        assert!(config.verbose);
    }
}

mod handle_failure_tests {
    use super::*;

    #[test]
    fn should_not_panic_when_disable_auto_issue_is_true() {
        let config = Config {
            disable_auto_issue: true,
            ..Config::default()
        };
        // Should return early without panicking
        handle_failure(&config, "test-cmd", "test-cmd arg1", 1, "/tmp/test.log");
    }

    #[test]
    fn should_not_panic_when_verbose_and_disable_auto_issue() {
        let config = Config {
            disable_auto_issue: true,
            verbose: true,
            ..Config::default()
        };
        handle_failure(&config, "test-cmd", "test-cmd --flag", 42, "/tmp/test.log");
    }
}

mod is_gh_authenticated_tests {
    use super::*;

    #[test]
    fn should_return_a_bool() {
        let result = is_gh_authenticated();
        // Just verify it returns a boolean without panicking
        let _ = result;
    }

    #[test]
    fn return_value_is_bool_type() {
        let result: bool = is_gh_authenticated();
        // Type assertion - if it compiles, it returns a bool
        let _ = result;
    }
}

mod is_gh_upload_log_available_tests {
    use super::*;

    #[test]
    fn should_return_a_bool() {
        let result = is_gh_upload_log_available();
        let _ = result;
    }

    #[test]
    fn return_value_is_bool_type() {
        let result: bool = is_gh_upload_log_available();
        let _ = result;
    }
}

mod can_create_issue_tests {
    use super::*;

    #[test]
    fn should_return_false_for_invalid_repo() {
        // A repo that doesn't exist should return false (gh not auth'd or repo not found)
        let result = can_create_issue("invalid-owner-xyz-99999", "invalid-repo-xyz-99999");
        assert!(!result);
    }

    #[test]
    fn should_return_bool() {
        let result: bool = can_create_issue("some-owner", "some-repo");
        let _ = result;
    }
}

/// Mirrors the JS `createIssue` regression tests for issue #168: the title and
/// body must reach `gh` verbatim, with real newlines and without shell escaping.
#[cfg(unix)]
mod create_issue_tests {
    use start_command::failure_handler::{create_issue, RepoInfo};
    use std::os::unix::fs::PermissionsExt;
    use std::process::Command;
    use tempfile::TempDir;

    const FAILING_COMMAND: &str = "echo \"quoted\" $(id) `hostname`";
    const HELPER_TEST: &str = "create_issue_tests::helper_process_calls_create_issue";
    const URL_PREFIX: &str = "created-issue-url:";

    /// Not a test on its own: re-executed by `create_issue_with_fake_gh` in a
    /// child process, because `PATH` has to be replaced only for that child.
    /// Mutating it in-process would also redirect the `gh` calls of the tests
    /// running next to it in the same binary.
    #[test]
    #[ignore = "helper process, driven by create_issue_with_fake_gh"]
    fn helper_process_calls_create_issue() {
        let repo_info = RepoInfo {
            owner: "owner".to_string(),
            repo: "repo".to_string(),
            url: "https://github.com/owner/repo".to_string(),
        };
        let url = create_issue(&repo_info, FAILING_COMMAND, 1, None).unwrap_or_default();
        println!("{}{}", URL_PREFIX, url);
    }

    /// Run `create_issue` with a fake `gh` first on the child's PATH that
    /// records its argv, NUL-separated, and return the URL and that argv.
    fn create_issue_with_fake_gh() -> (Option<String>, Vec<String>) {
        let dir = TempDir::new().unwrap();
        let argv_file = dir.path().join("argv");
        let gh_path = dir.path().join("gh");
        std::fs::write(
            &gh_path,
            format!(
                "#!/bin/sh\nprintf '%s\\0' \"$@\" > \"{}\"\necho https://github.com/owner/repo/issues/1\n",
                argv_file.display()
            ),
        )
        .unwrap();
        let mut permissions = std::fs::metadata(&gh_path).unwrap().permissions();
        permissions.set_mode(0o755);
        std::fs::set_permissions(&gh_path, permissions).unwrap();

        let mut paths = vec![dir.path().to_path_buf()];
        if let Some(existing) = std::env::var_os("PATH") {
            paths.extend(std::env::split_paths(&existing));
        }

        let output = Command::new(std::env::current_exe().unwrap())
            .args(["--exact", HELPER_TEST, "--ignored", "--nocapture"])
            .env("PATH", std::env::join_paths(paths).unwrap())
            .output()
            .unwrap();
        assert!(
            output.status.success(),
            "helper process failed: {}",
            String::from_utf8_lossy(&output.stderr)
        );

        let stdout = String::from_utf8_lossy(&output.stdout).to_string();
        let url = stdout
            .lines()
            .find_map(|line| line.strip_prefix(URL_PREFIX))
            .filter(|url| !url.is_empty())
            .map(str::to_string);

        let recorded = std::fs::read_to_string(&argv_file).unwrap();
        // `printf '%s\0'` writes a trailing NUL, so the final split part is empty.
        let mut argv = recorded.split('\0').map(str::to_string).collect::<Vec<_>>();
        argv.pop();
        (url, argv)
    }

    #[test]
    fn passes_the_title_and_body_as_separate_arguments() {
        let (url, argv) = create_issue_with_fake_gh();
        assert_eq!(
            url.as_deref(),
            Some("https://github.com/owner/repo/issues/1")
        );
        assert_eq!(argv[0..4], ["issue", "create", "--repo", "owner/repo"]);
        assert_eq!(argv[4], "--title");
        assert_eq!(argv[6], "--body");
        assert_eq!(argv.len(), 8);
    }

    #[test]
    fn keeps_the_failing_command_verbatim_and_writes_real_newlines() {
        let (_, argv) = create_issue_with_fake_gh();
        let title = &argv[5];
        let body = &argv[7];
        assert!(
            title.contains(FAILING_COMMAND),
            "title must quote the command verbatim, got: {}",
            title
        );
        assert!(
            body.contains(FAILING_COMMAND),
            "body must quote the command verbatim"
        );
        assert!(!title.contains("\\\""), "title must not escape quotes");
        assert!(!body.contains("\\\""), "body must not escape quotes");
        assert!(body.contains('\n'), "body must contain real newlines");
        assert!(
            !body.contains("\\n"),
            "body must not contain literal backslash-n sequences"
        );
    }
}

File: js/test/failure-handler.js
#!/usr/bin/env bun
/**
 * Unit tests for failure-handler module
 * Tests pure functions: parseGitUrl and handleFailure early-exit behavior
 */

const { describe, it } = require('node:test');
const assert = require('assert');
const fs = require('fs');
const os = require('os');
const path = require('path');
const { execFileSync } = require('child_process');
const { parseGitUrl, handleFailure } = require('../src/lib/failure-handler');

/**
 * Run `createIssue` in a child process whose PATH starts with a fake `gh`
 * recording its argv (NUL-separated). A child process is required because the
 * runtime resolves PATH at startup, so mutating `process.env.PATH` in-process
 * does not affect command lookup under Bun.
 */
function createIssueWithFakeGh(fullCommand) {
  const dir = fs.mkdtempSync(path.join(os.tmpdir(), 'fake-gh-'));
  try {
    const argvFile = path.join(dir, 'argv');
    const ghPath = path.join(dir, 'gh');
    fs.writeFileSync(
      ghPath,
      [
        '#!/bin/sh',
        `printf '%s\\0' "$@" > "${argvFile}"`,
        'echo https://github.com/owner/repo/issues/1',
        '',
      ].join('\n')
    );
    fs.chmodSync(ghPath, 0o755);

    const modulePath = require.resolve('../src/lib/failure-handler');
    const script = `
      const { createIssue } = require(${JSON.stringify(modulePath)});
      const repo = { owner: 'owner', repo: 'repo', url: 'https://github.com/owner/repo' };
      process.stdout.write(
        String(createIssue(repo, ${JSON.stringify(fullCommand)}, 1, null))
      );
    `;
    const result = execFileSync(process.execPath, ['-e', script], {
      encoding: 'utf8',
      env: {
        ...process.env,
        PATH: `${dir}${path.delimiter}${process.env.PATH}`,
      },
    });
    const argv = fs.readFileSync(argvFile, 'utf8').split('\0').slice(0, -1);
    return { result, argv };
  } finally {
    fs.rmSync(dir, { recursive: true, force: true });
  }
}

describe('failure-handler', () => {
  describe('parseGitUrl', () => {
    it('should parse HTTPS GitHub URL', () => {
      const result = parseGitUrl('https://github.com/owner/my-repo');
      assert.ok(result !== null);
      assert.strictEqual(result.owner, 'owner');
      assert.strictEqual(result.repo, 'my-repo');
      assert.strictEqual(result.url, 'https://github.com/owner/my-repo');
    });

    it('should parse HTTPS URL with .git suffix', () => {
      const result = parseGitUrl('https://github.com/owner/my-repo.git');
      assert.ok(result !== null);
      assert.strictEqual(result.owner, 'owner');
      assert.strictEqual(result.repo, 'my-repo');
      assert.strictEqual(result.url, 'https://github.com/owner/my-repo');
    });

    it('should parse SSH git@ URL', () => {
      const result = parseGitUrl('git@github.com:owner/my-repo.git');
      assert.ok(result !== null);
      assert.strictEqual(result.owner, 'owner');
      assert.strictEqual(result.repo, 'my-repo');
    });

    it('should parse git+https URL format', () => {
      const result = parseGitUrl('git+https://github.com/owner/repo.git');
      assert.ok(result !== null);
      assert.strictEqual(result.owner, 'owner');
      assert.strictEqual(result.repo, 'repo');
    });

    it('should return null for empty string', () => {
      const result = parseGitUrl('');
      assert.strictEqual(result, null);
    });

    it('should return null for null/undefined input', () => {
      assert.strictEqual(parseGitUrl(null), null);
      assert.strictEqual(parseGitUrl(undefined), null);
    });

    it('should return null for non-github URL', () => {
      const result = parseGitUrl('https://gitlab.com/owner/repo');
      assert.strictEqual(result, null);
    });

    it('should return null for invalid/random string', () => {
      const result = parseGitUrl('not-a-url-at-all');
      assert.strictEqual(result, null);
    });

    it('should normalize URL to https://github.com format', () => {
      const result = parseGitUrl('git@github.com:myorg/myrepo');
      assert.ok(result !== null);
      assert.ok(result.url.startsWith('https://github.com/'));
    });

    it('should handle URL with subdirectory (only owner/repo captured)', () => {
      const result = parseGitUrl('https://github.com/myorg/myrepo/issues');
      assert.ok(result !== null);
      assert.strictEqual(result.owner, 'myorg');
      assert.strictEqual(result.repo, 'myrepo');
    });

    it('should return object with owner, repo, url keys', () => {
      const result = parseGitUrl('https://github.com/test/project');
      assert.ok(result !== null);
      assert.ok(Object.prototype.hasOwnProperty.call(result, 'owner'));
      assert.ok(Object.prototype.hasOwnProperty.call(result, 'repo'));
      assert.ok(Object.prototype.hasOwnProperty.call(result, 'url'));
    });
  });

  describe('createIssue', () => {
    // Regression for issue #168: the command was interpolated into a shell
    // string and only `"` was escaped, so a backtick or $(...) in the failing
    // command ran, and every newline of the body arrived as a literal "\\n".
    const runCreateIssue = () =>
      createIssueWithFakeGh('echo "quoted" $(id) `hostname`');

    it('passes the title and body to gh as separate arguments', () => {
      if (process.platform === 'win32') {
        return; // /bin/sh stub is POSIX-only
      }
      const { result, argv } = runCreateIssue();
      assert.strictEqual(result, 'https://github.com/owner/repo/issues/1');
      assert.deepStrictEqual(argv.slice(0, 4), [
        'issue',
        'create',
        '--repo',
        'owner/repo',
      ]);
      assert.strictEqual(argv[4], '--title');
      assert.strictEqual(argv[6], '--body');
      assert.strictEqual(argv.length, 8);
    });

    it('keeps the failing command verbatim and does not escape it', () => {
      if (process.platform === 'win32') {
        return;
      }
      const { argv } = runCreateIssue();
      const [title, body] = [argv[5], argv[7]];
      assert.ok(
        title.includes('echo "quoted" $(id) `hostname`'),
        `title must quote the command verbatim, got: ${title}`
      );
      assert.ok(
        body.includes('echo "quoted" $(id) `hostname`'),
        'body must quote the command verbatim'
      );
      assert.ok(
        !title.includes('\\"'),
        'title must not contain escaped quotes'
      );
      assert.ok(!body.includes('\\"'), 'body must not contain escaped quotes');
    });

    it('writes real newlines into the issue body', () => {
      if (process.platform === 'win32') {
        return;
      }
      const { argv } = runCreateIssue();
      const body = argv[7];
      assert.ok(body.includes('\n'), 'body must contain real newlines');
      assert.ok(
        !body.includes('\\n'),
        'body must not contain literal backslash-n sequences'
      );
    });
  });

  describe('handleFailure', () => {
    it('should return early when disableAutoIssue is true', () => {
      // This should not throw and should return without calling external processes
      const config = { disableAutoIssue: true };
      // If it tries to call external tools, it would either throw or hang;
      // returning cleanly means the early-exit path was taken.
      assert.doesNotThrow(() => {
        handleFailure(config, 'someCmd', 'someCmd --flag', 1, '/tmp/fake.log');
      });
    });

    it('should return early when disableAutoIssue is true (verbose mode)', () => {
      const config = { disableAutoIssue: true, verbose: true };
      assert.doesNotThrow(() => {
        handleFailure(config, 'cmd', 'cmd arg', 2, '/tmp/fake.log');
      });
    });
  });
});
